### ISSUE TITLE
Custom style functions

### ISSUE THREAD

**Issue by paul121** [farmOS/farmOS-map#72]
I'd like to be able to pass in a custom Style Function that styles features based on feature properties..

A simple example for a `geojson` layer:
```js
        function myStyleFunction(feature, resolution) {
            var areaID = +feature.get('id');

            var defaultColor = 'rgba(0, 0, 0, 1)';
            var color1 = 'rgba(100, 100, 100, 1)';
            var color2 = 'rgba(200, 200, 200, 1)';

            let color = defaultColor;
            if (areaID < 10) {
                color = color1
            }
            if (areaID >= 10) {
                color = color2
            }

            return new ol.style.Style({
                fill: new ol.style.Fill({
                    color
                }),
                stroke: new ol.style.Stroke({
                    color: [220,220,220,1],
                    width: 1
                })
            });
        }

        var opts = {
            title: 'Title',
            url,
            styleFunction: myStyleFunction,
        }
        instance.addLayer('geojson', opts);
```

I've tested the following modification of `addGeoJSONLayer` to add this functionality and it seems to work well:
```js
// Add a GeoJSON feature layer to the map.
function addGeoJSONLayer({
  title = 'geojson', url, color = 'orange', styleFunction = null, visible = true,
}) {
  const format = new GeoJSON();
  const source = new VectorSource({ url, format });
  const layer = new VectorLayer({
    title,
    source,
    style: styleFunction ? styleFunction : styles(color),
    visible,
  });
  return layer;
}
```

I don't this would require adding any additional dependencies to the farmOS-map library. Perhaps it would be good to check if the `styleFunction` was "valid" (per [StyleFunction docs](https://openlayers.org/en/latest/apidoc/module-ol_style_Style.html#~StyleFunction)), but are there any considerations here? I imagine this could apply to other types of `TileLayer`s, too, although I'm not super familiar.

Motivation for this comes from wanting to style features queried via `/farm/areas/geojson/`, _but without_ the need to create individual Features from farmOS area geometries in a custom farmOS-map behavior.

**Comment by mstenta** [comment]
Cool idea! I'm open to adding this. I'd like to do it in a consistent way with other layer types, if they support it.

It also makes me wonder about refactoring `layer.js` to better handle common parameters. I feel like that file is starting to get harder to maintain. But that's beside the point of this issue - just thinking out loud. :-)

Another one of the challenges we face more generally with this library, which might be relevant here, is that it includes all the OpenLayers modules it uses, but doesn't expose them for other code to use.

So, for example... you're using `ol.style.Style`, `ol.style.Fill`, and `ol.style.Stroke` in your style function. Did you have to include the OpenLayers library itself alongside the farmOS-map.js to get that to work? I assume so. That means you have OpenLayers code *twice* in your page, which is a big waste. But that's the only way to do it, based on the way farmOS-map.js packages things right now. We may want to think about that more in a separate issue. It's not a blocker for this, technically, but certainly affects anyone who wants to create their own style function.

**Comment by mstenta** [comment]
> Did you have to include the OpenLayers library itself alongside the farmOS-map.js to get that to work?

Of course... if you are bundling your custom JS code and farmOS-map.js together into a custom package with Webpack, then this isn't an issue. But @jgaehring has a better understanding of those considerations than I do - he's bundling farmOS-map into farmOS-client via Webpack.

Related: https://github.com/farmOS/farmOS-map/issues/68

**Comment by jgaehring** [comment]
Maybe `ol.style` could be provided as a third parameter to the style function?

```js
        function myStyleFunction(feature, resolution, style) {
            var areaID = +feature.get('id');

            var defaultColor = 'rgba(0, 0, 0, 1)';
            var color1 = 'rgba(100, 100, 100, 1)';
            var color2 = 'rgba(200, 200, 200, 1)';

            let color = defaultColor;
            if (areaID < 10) {
                color = color1
            }
            if (areaID >= 10) {
                color = color2
            }

            return new style.Style({
                fill: new style.Fill({
                    color
                }),
                stroke: new style.Stroke({
                    color: [220,220,220,1],
                    width: 1
                })
            });
        }
```

then...

```js
// Add a GeoJSON feature layer to the map.
function addGeoJSONLayer({
  title = 'geojson', url, color = 'orange', styleFunction = null, visible = true,
}) {
  const format = new GeoJSON();
  const source = new VectorSource({ url, format });
  const style = styleFunction
    ? (feature, resolution) => styleFunction(feature, resolution, ol.style)
    : styles(color);
  const layer = new VectorLayer({
    title,
    source,
    style,
    visible,
  });
  return layer;
}
```

**Comment by mstenta** [comment]
I like where you're going with this @jgaehring !

But: we don't have `ol.style` either. Here is how we are using styles currently: https://github.com/farmOS/farmOS-map/blob/master/src/styles/index.js

But I suppose we could create our own `style` object that includes `Style`, `Stroke`, `Fill`, etc?

**Comment by mstenta** [comment]
Maybe we want to rethink `src/styles/index.js` more broadly? And work the custom style function capabilities into that file somehow?

**Comment by jgaehring** [comment]
> But: we don't have ol.style either.

If you import `ol.stye` into [`/instance/methods/layer.js`](https://github.com/farmOS/farmOS-map/blob/master/src/instance/methods/layer.js) it won't result in duplicate code in the farmOS-map bundle. Those dependencies can be analyzed by Webpack at compile-time and optimized so they're both using the same reference. The only problem is with code that uses the global, because Webpack can't do runtime optimizations like that.

**Comment by paul121** [comment]
> Did you have to include the OpenLayers library itself alongside the farmOS-map.js to get that to work? I assume so. That means you have OpenLayers code twice in your page, which is a big waste.

Yea, I'm doing this in the early stages for now :-) I wasn't sure if `ol.style` could be provided by `farmOS-map` or not. It certainly isn't a requirement for what I'm doing, but agree that including this could benefit many other use cases.

Is something like `import {ol.style as style} from farmOS-map` possible?

**Comment by jgaehring** [comment]
> Is something like import {ol.style as style} from farmOS-map possible?

You can just do

```js
import style from 'ol/style';
```

then use `new style.Fill` or what have you.

That will give you an object literal containing all the classes: https://github.com/openlayers/openlayers/blob/master/src/ol/style.js

**Comment by paul121** [comment]
@jgaehring right, I've got that.. but in the context of _using_ farmOS-map does that work?

Right now `ol/style` isn't included in farmOS-map. But if it were, does that mean `ol/style` is now available globally?

From my example above, could I do: `import style from 'ol/style';` in my JS file that's included in `static/index.html` ?

**Comment by jgaehring** [comment]
Oh, gotcha. That's essentially equivalent to importing from `node_modules/ol/style`, which is essentially the same as if you were to install OL via npm. Although if I had to guess, I'd say Webpack will probably no recognize `'ol/style'` if OL is not in your `package.json`. So you could try it, it might work, but even if it did, I would not recommend it. You're better off just doing it the proper way and `npm install`ing it.

Again, I think the best route here is to pass `ol.style` as a parameter to `styleFunction`; as I mentioned, this won't have adverse effects on the bundle size of farmOS-map or your application.

**Comment by paul121** [comment]
Interesting, thanks! Passing `ol.style` as a parameter makes sense. I might hack on some of that.

I guess I was hoping `ol.style` could be provided in a more "global" way by farmOS-map, but maybe that's not possible. I'm also curious in @mstenta idea of making `src/styles` more broad.. interested if that could achieve something similar. Will need to think on it more...

**Comment by jgaehring** [comment]
Hm, I think there are problems with the "global" approach which has partly led us to this conundrum, so I'd caution against taking parts of the OL API and wholesale attaching them onto the global.

In an ideal world, there would be no need for the global, and farmOS-map could just import `ol.style`, the application code could import farmOS-map, and if the application code also wanted to import `ol.style` separately, Webpack would recognize that as the same dependency and wouldn't bundle it twice. I think that's the path forward, where we want to get to. If Drupal needs a global, that's fine, hopefully we can find a way to add that on after the fact, but I think overloading that global is fraught with more issues than its worth.

**Comment by paul121** [comment]
Ah yes! For application code that imports farmOS-map, I agree there is no need for a global. In those cases it should be easy enough to import `ol.style` separately.

Sorry, I should have explained myself better... I am thinking in terms of the Drupal use case :-)

More specifically, using styles from within a farmOS-map behavior (without any other dependencies). The example I mentioned above would use a `styleFunction` to style features based on their custom properties. Including `ol.style` in the parameter for `styleFunction` would work well for that.

But I am curious about Drupal use cases that would need `ol/style` from outside of the `styleFunction`. In general, adding a layer with styled features directly to the map (a good example [here](https://openlayers.org/en/latest/examples/regularshape.html?q=style)) would require this. This might be useful for displaying icons, symbols or other features on the map. This seems useful.. but I guess the question might be: can these use cases be supported by farmOS-map? Or do they warrant needing to build another JS module just to provide the required `ol` attributes?

**Comment by jgaehring** [comment]
> But I am curious about Drupal use cases that would need ol/style from outside of the styleFunction. In general, adding a layer with styled features directly to the map (a good example here) would require this.

Hmm, yea, again, wish I understood how Drupal was using that global. I keep coming back to issue #68, and feel there has to be some better way to handle that. Maybe sometime we could look at that all together. Perhaps some form of dependency injection could work? So in Drupal land the OL deps could get injected from the global, and from Webpack land they get injected from a plain ES6 import? I dunno. :thinking:

**Comment by mstenta** [comment]
> Again, I think the best route here is to pass `ol.style` as a parameter to `styleFunction`; as I mentioned, this won't have adverse effects on the bundle size of farmOS-map or your application.

Agreed! This seems like the best approach.

> I'm also curious in @mstenta idea of making `src/styles` more broad.. interested if that could achieve something similar. Will need to think on it more...

I don't have any specific thoughts in mind... just thinking that `src/styles` is where I put all the style-related code, so if it makes sense to put anything else in there for organizational purposes that would be nice. But simply passing `ol.style` as a parameter to `styleFunction` could just be done in `layers.js` - can't really think of any way to break that out without making things more complex than they need to be.

> Hm, I think there are problems with the "global" approach which has partly led us to this conundrum, so I'd caution against taking parts of the OL API and wholesale attaching them onto the global.
> The only problem is with code that uses the global, because Webpack can't do runtime optimizations like that.
> Hmm, yea, again, wish I understood how Drupal was using that global. I keep coming back to issue #68, and feel there has to be some better way to handle that. Maybe sometime we could look at that all together. Perhaps some form of dependency injection could work? So in Drupal land the OL deps could get injected from the global, and from Webpack land they get injected from a plain ES6 import? I dunno. thinking

Regarding all of @jgaehring's points about the global `window.farmOS.map` and adding pieces of OL to it: I agree 100%.

I'm sure we can figure out a way to remove/disable that global when you just want to bundle farmOS-map yourself.

We should probably keep this conversation focused over in #68 though. I'll post a comment over there that describes a bit more how the global is being used in the Drupal context and why.

**Comment by paul121** [comment]
Thanks @mstenta that is great info about the global in Drupal. Step by step learning how this all fits together 😃 

I'll work on a PR for a `styleFunction` following @jgaehring idea of passing `ol.style` as a parameter. This will be quite useful, I'm glad there is a simple solution!

**Comment by mstenta** [comment]
Merged PR #78
